### ISSUE TITLE
FSDP Hybrid Shard worse loss than than Full Shard

### ISSUE THREAD

**Issue by zaptrem** [pytorch/pytorch#140183]
### 🐛 Describe the bug

When using HYBRID_SHARD instead of FULL_SHARD on PyTorch 2.4.1 the loss of our model behaves similarly to when it is being trained on one node (despite training on two). When using FULL_SHARD, the loss behaves as expected when training on two nodes (so twice the effective batch size). Additionally, the problem also happens with FULL_SHARD when on PyTorch 2.5. See the screenshot below 
![image](https://github.com/user-attachments/assets/8c8ab8bf-7288-4784-a351-6152df42fe8c)

The device meshes we use are as follows:

```
        if self.strategy.sharding_strategy in [ShardingStrategy._HYBRID_SHARD_ZERO2, ShardingStrategy.HYBRID_SHARD]:
            self.device_mesh = init_device_mesh("cuda", (world_size//8, 8), mesh_dim_names=("replicate", "shard"))
        else:
            self.device_mesh = init_device_mesh("cuda", (world_size,))
```

### Versions

https://gist.github.com/zaptrem/799d0c99c9e69067eb937ab1a55f1c69

**Comment by awgu** [comment]
Are you using `torch.compile`?

**Comment by zaptrem** [comment]
> Are you using `torch.compile`?

Yes

**Comment by awgu** [comment]
Are you able to try to repro the issue without `torch.compile`? Between 2.4 and 2.5, there were no real changes to FSDP, as we have been working on FSDP2. I suspect there might have been some change on the compile side though. cc: @bdhirsh

**Comment by mtywh** [comment]
In my model training experiments, I also observed accuracy degradation when applying FSDP Hybrid Sharding on a 4-node, 32-GPU setup. I ran ablation experiments with different mesh=(dp, fsdp) configurations, and the results are as follows:

FULL_SHARD: AUC = 0.8189
HYBRID_SHARD + mesh=(1, 16): AUC = 0.8190
HYBRID_SHARD + mesh=(2, 8): AUC = 0.8174
HYBRID_SHARD + mesh=(4, 4): AUC = 0.8155
HYBRID_SHARD + mesh=(16, 1): AUC = 0.8134


I ruled out torch.compile as the cause. The root cause turned out to be the lack of random seed synchronization across DP groups during model parameter initialization. Because different DP groups started with inconsistent model parameters, the gradient all-reduce effectively mixed gradients from divergent models, pulling the optimization away from the correct descent direction and ultimately converging to a suboptimal solution.

In my case, the fix was to have node 0 determine a seed value and broadcast it to all other nodes, then call torch.manual_seed() to fix the random seed before model initialization. After this change, the accuracy of HYBRID_SHARD recovered to the same level as FULL_SHARD.

HYBRID_SHARD + mesh=(16, 1) + random_seed_fix: AUC = 0.8189
HYBRID_SHARD + mesh=(2, 8) + random_seed_fix: AUC = 0.8192

cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @weifengpy @zhaojuanmao @mrshenli @rohan-varma @chauhang @mori360 @ppwwyyxx @H-Huang @kwen2501 @c-p-i-o

**Comment by weifengpy** [comment]
@mtywh @zaptrem thanks for the issue and diagnose.  FSDP1 has `sync_module_states=True` to broadcasts parameters and buffers, preventing ranks starting with divergent models due to different seed

```
FSDP(
      model,
      device_mesh=device_mesh,
      sharding_strategy=ShardingStrategy.HYBRID_SHARD,
      sync_module_states=True,   # <----------------- turn it on
      device_id=local_rank,
      use_orig_params=True,
  )
```

**Comment by mtywh** [comment]
Thanks for the attention and feedback！
One thing I forgot to mention is that I'm using FSDP2. As far as I know, in FSDP2 the sync_module_states functionality has been moved to DCP, which makes this setting fairly easy to miss. 
So I followed the approach used in [TorchTitan](https://github.com/pytorch/torchtitan), which synchronizes model parameters by broadcasting the random seed before initialization.

> [@mtywh](https://github.com/mtywh) [@zaptrem](https://github.com/zaptrem) thanks for the issue and diagnose. FSDP1 has `sync_module_states=True` to broadcasts parameters and buffers, preventing ranks starting with divergent models due to different seed
> 
> ```
> FSDP(
>       model,
>       device_mesh=device_mesh,
>       sharding_strategy=ShardingStrategy.HYBRID_SHARD,
>       sync_module_states=True,   # <----------------- turn it on
>       device_id=local_rank,
>       use_orig_params=True,
>   )
> ```

**Comment by weifengpy** [comment]
> Thanks for the attention and feedback！ One thing I forgot to mention is that I'm using FSDP2. As far as I know, in FSDP2 the sync_module_states functionality has been moved to DCP, which makes this setting fairly easy to miss. So I followed the approach used in [TorchTitan](https://github.com/pytorch/torchtitan), which synchronizes model parameters by broadcasting the random seed before initialization.

@mtywh are you using huggingface or some trainer? I noticed you used term "HYBRID_SHARD" and "FULL_SHARD" whcih are fsdp1 contants

for FSDP2, we don't have sync_module_states because DTensor handle random ops by broadcasting seed from rank0 to other ranks. Did you mention DCP because of loading from a checkpoint? (but also loading checkpoint means no need for random init)

**Comment by mtywh** [comment]
@weifengpy To address a few questions:

1. I am using a self-built training framework based on torch and torchrec, with FSDP2 integrated into it.

2. I use the terms FULL_SHARD and HYBRID_SHARD out of habit, since FSDP2 can represent the sharding strategies from FSDP1 through the device mesh configuration and the behavior of reshard_after_forward.

3. In the experiments mentioned above, I did not observe DTensor-related handling that automatically synchronizes the model state. I could only explicitly set the seed via torch.manual_seed(). Which setting should be used to enable this behavior?

4. I mentioned DCP because I saw in [FSDP_tutorial](https://docs.pytorch.ac.cn/tutorials/intermediate/FSDP_tutorial.html) that the behavior of sync_module_states (FSDP1) has been moved to DCP’s set_model_state_dict(broadcast_from_rank0=True). In my code, training from scratch requires random initialization, while incremental training can use DCP to control and synchronize the model state.


> > Thanks for the attention and feedback！ One thing I forgot to mention is that I'm using FSDP2. As far as I know, in FSDP2 the sync_module_states functionality has been moved to DCP, which makes this setting fairly easy to miss. So I followed the approach used in [TorchTitan](https://github.com/pytorch/torchtitan), which synchronizes model parameters by broadcasting the random seed before initialization.
> 
> [@mtywh](https://github.com/mtywh) are you using huggingface or some trainer? I noticed you used term "HYBRID_SHARD" and "FULL_SHARD" whcih are fsdp1 contants
> 
> for FSDP2, we don't have sync_module_states because DTensor handle random ops by broadcasting seed from rank0 to other ranks. Did you mention DCP because of loading from a checkpoint? (but also loading checkpoint means no need for random init)
